packages/langium/test/workspace/document-builder.test.ts

```diff
@@ -238,6 +239,88 @@ describe('DefaultDocumentBuilder', () => {
+            expected.push('B' + state);
+            expected.push('W' + state);
+            builder.onBuildPhase(state, async () => {
+                actual.push('B' + state);
+                await delayNextTick();
+            });
+            builder.waitUntil(state).then(() => actual.push('W' + state));
+        }
+        for (let i = 2; i <= 6; i++) {
+            wait(i);
+        }
+        await builder.build([document], { validation: true });
+        expect(actual).toEqual(expected);
+    });
+
+    test('`waitUntil` will correctly wait even though the build process has been cancelled', async () => {
+        const services = await createServices();
+        const documentFactory = services.shared.workspace.LangiumDocumentFactory;
+        const documents = services.shared.workspace.LangiumDocuments;
+        const builder = services.shared.workspace.DocumentBuilder;
+        const document = documentFactory.fromString('', URI.parse('file:///test1.txt'));
+        documents.addDocument(document);
+
+        const actual: string[] = [];
+        const cancelTokenSource = new CancellationTokenSource();
+        function wait(state: DocumentState): void {
+            builder.onBuildPhase(state, async () => {
+                actual.push('B' + state);
+                await delayNextTick();
+            });
+        }
+        for (let i = 2; i <= 6; i++) {
+            wait(i);
+        }
+        builder.waitUntil(DocumentState.ComputedScopes).then(() => cancelTokenSource.cancel());
+        builder.waitUntil(DocumentState.IndexedReferences).then(() => {
+            actual.push('W' + DocumentState.IndexedReferences);
+        });
+        // Build twice but interrupt the first build after the computing scope phase
+        try {
+            await builder.build([document], { validation: true }, cancelTokenSource.token);
+        } catch {
+            // build has been cancelled, ignore
+        }
+        document.state = DocumentState.Parsed;
+        await builder.build([document], { validation: true });
+        // The B2 and B3 phases are duplicated because the first build has been cancelled
+        // W5 still appears as expected after B5
+        expect(actual).toEqual(['B2', 'B3', 'B2', 'B3', 'B4', 'B5', 'W5', 'B6']);
+    });
+
+    test('`waitUntil` can be cancelled before it gets triggered', async () => {
+        const services = await createServices();
+        const documentFactory = services.shared.workspace.LangiumDocumentFactory;
+        const documents = services.shared.workspace.LangiumDocuments;
+        const builder = services.shared.workspace.DocumentBuilder;
+        const document = documentFactory.fromString('', URI.parse('file:///test1.txt'));
+        documents.addDocument(document);
+
+        const cancelTokenSource = new CancellationTokenSource();
+        builder.waitUntil(DocumentState.IndexedReferences, cancelTokenSource.token).then(() => {
+            fail('This should have been cancelled');
```

Comment:
So, this is assumed to be never reached? Please leave a comment, or make that explicit in the message.